internal/runner/runnerpool/runner.go

```diff
@@ -0,0 +1,318 @@
+		}
+
+		return res
+	}
+
+	pool := New(r.Stack.Units, taskRun, opts.Parallelism, opts.FailFast)
+
+	results := pool.Run(ctx, l)
+
+	var allErrs []error
+
+	for _, res := range results {
+		if res.Err != nil {
+			allErrs = append(allErrs, res.Err)
+		}
+	}
+
+	if len(allErrs) > 0 {
+		return errors.Join(allErrs...)
+	}
+
+	return nil
+}
+
+// handleApplyDestroy handles logic for apply and destroy commands.
+func (r *Runner) handleApplyDestroy(l log.Logger, opts *options.TerragruntOptions) {
+	if opts.RunAllAutoApprove {
+		opts.TerraformCliArgs = util.StringListInsert(opts.TerraformCliArgs, "-auto-approve", 1)
+	}
+
+	r.syncTerraformCliArgs(l, opts)
+}
+
+// handleShow handles logic for show command.
+func (r *Runner) handleShow(l log.Logger, opts *options.TerragruntOptions) {
+	r.syncTerraformCliArgs(l, opts)
+}
+
+// handlePlan handles logic for plan command, including error buffer setup and summary.
+func (r *Runner) handlePlan() {
+	r.planErrorBuffers = make([]bytes.Buffer, len(r.Stack.Units))
+	for i, u := range r.Stack.Units {
+		u.TerragruntOptions.ErrWriter = io.MultiWriter(&r.planErrorBuffers[i], u.TerragruntOptions.ErrWriter)
+	}
+}
+
+func (r *Runner) LogUnitDeployOrder(l log.Logger, terraformCommand string) error {
+	outStr := fmt.Sprintf("The runner-pool runner at %s will be processed in the following order for command %s:\n", r.Stack.TerragruntOptions.WorkingDir, terraformCommand)
+	for _, unit := range r.Stack.Units {
+		outStr += fmt.Sprintf("Unit %s\n", unit.Path)
+	}
+
+	l.Info(outStr)
+
+	return nil
+}
+
+func (r *Runner) JSONUnitDeployOrder(terraformCommand string) (string, error) {
+	orderedUnits := make([]string, 0, len(r.Stack.Units))
+	for _, unit := range r.Stack.Units {
+		orderedUnits = append(orderedUnits, unit.Path)
+	}
+
+	j, err := json.MarshalIndent(orderedUnits, "", "  ")
+	if err != nil {
+		return "", err
+	}
+
+	return string(j), nil
+}
+
+func (r *Runner) ListStackDependentUnits() map[string][]string {
+	dependentUnits := make(map[string][]string)
+
+	for _, unit := range r.Stack.Units {
+		if len(unit.Dependencies) != 0 {
+			for _, dep := range unit.Dependencies {
+				dependentUnits[dep.Path] = util.RemoveDuplicatesFromList(append(dependentUnits[dep.Path], unit.Path))
+			}
+		}
+	}
+
+	for {
+		noUpdates := true
+
+		for unit, dependents := range dependentUnits {
+			for _, dependent := range dependents {
+				initialSize := len(dependentUnits[unit])
+				list := util.RemoveDuplicatesFromList(append(dependentUnits[unit], dependentUnits[dependent]...))
+				list = util.RemoveElementFromList(list, unit)
+				dependentUnits[unit] = list
+
+				if initialSize != len(dependentUnits[unit]) {
+					noUpdates = false
+				}
+			}
+		}
+
+		if noUpdates {
+			break
+		}
+	}
+
+	return dependentUnits
+}
+
+// Sync the TerraformCliArgs for each unit in the stack to match the provided terragruntOptions struct.
+func (r *Runner) syncTerraformCliArgs(l log.Logger, opts *options.TerragruntOptions) {
```

Comment:
Is there any way we can avoid increasing reliance on TerragruntOptions here? Totally fine if the answer is no, or no, we should tackle that later.